Introduce a `unit_test` build tag for unit test-only files

timstclair opened the pull request:
Adds a `test` build tag for files which should only be included in testing builds. This is to prevent leaking test resources (such as the `httptest.serve` flag) while avoiding the need for tricky refactorings (e.g. moving these files into testing packages, which can result in circlar dependencies that are messy to resolve). The downside of this change is that running unit tests always requires the `-tags unit_test` flag now (it's too bad go doesn't have an automatic test-only flag).

Fixes https://github.com/kubernetes/kubernetes/issues/21114

ixdy commented on hack/test-go.sh:
```diff
@@ -121,7 +121,7 @@ done
 shift $((OPTIND - 1))
 
 # Use eval to preserve embedded quoted strings.
-eval "goflags=(${KUBE_GOFLAGS:-})"
```
maybe always include `-tags test`? or does `go build` not allow `-tags` to be specified multiple times?

in particular, `hack/test-integration.sh` (which calls `hack/test-go.sh`) sets `KUBE_GOFLAGS`.

timstclair replied:
I think it's debateable whether the integration tests should set the test build tag... I see that mesos uses the `unit_test` tag, which may be better in this case... WDYT?

ixdy replied:
yeah, `unit_test` might be better, to differentiate from things used by integration or e2e tests.

timstclair replied:
Done.

tallclair committed:
Introduce a `test` build tag for test-only files

Adds a `test` build tag for files which should only be included in
testing builds. This is to prevent leaking test resources (such as the
`httptest.serve` flag) while avoiding the need for tricky
refactorings (e.g. moving these files into testing packages, which can
result in circlar dependencies that are messy to resolve). The downside
of this change is that running unit tests always requires the `-tags
test` flag now.

thockin commented:
Nice!!!

ixdy commented:
The integration test is failing to build:
```
# k8s.io/kubernetes/test/integration
_output/local/go/src/k8s.io/kubernetes/test/integration/persistent_volumes_test.go:52: undefined: volume.NewFakeVolumeHost
_output/local/go/src/k8s.io/kubernetes/test/integration/persistent_volumes_test.go:54: undefined: volume.FakeVolumePlugin
FAIL	k8s.io/kubernetes/test/integration [build failed]
```

timstclair commented:
From the discussion on https://github.com/kubernetes/kubernetes/issues/21114 the prefered fix is to refactor the code rather than add build tags, so I'm closing this.
